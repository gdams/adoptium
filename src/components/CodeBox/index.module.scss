@import "./light.module.scss";

%base-styles {
  border-radius: 0.3rem;
  font-family: Consolas, Monaco, "Andale Mono", "Ubuntu Mono", monospace;
  font-size: 1em;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
  line-height: 1.5;
  margin: 0.5em 0 var(--space-24) 0;
  overflow: auto;
  padding: 0;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;
  text-align: left;
  white-space: pre;
  word-break: normal;
  word-spacing: normal;
  word-wrap: normal;
}

.code {
  @extend %base-styles;
  font-weight: var(--font-weight-light);
  padding: 0 6px;

  white-space: break-spaces;
}

.pre {
  @extend %base-styles;

  padding: 0 !important;

  .top {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    span,
    button {
      align-items: center;
      display: inherit;
      font-size: var(--font-size-code);
      height: 23px;
      justify-content: center;
      width: 86px;
    }

    span {
      background-color: var(--black4);
      border-radius: 0 0 0.3rem;
      color: var(--black9);
    }

    button {
      background-color: var(--black9);
      border-radius: 0 0 0 0.3rem;
      border-width: 0;
      color: white;

      i {
        padding: 0;
      }
    }
  }

  .content {
    margin: 1em;
  }
}

:global(.light) {
  .pre {
    @include light-styles;
  }

  .code {
    background-color: var(--black2);
    color: var(--black9);
  }

  a .code {
    background-color: transparent;
    color: var(--color-text-accent);
    padding: 0;
  }
}
