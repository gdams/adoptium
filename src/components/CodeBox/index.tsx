import React, { useEffect, useState } from 'react';
import { highlight, languages } from 'prismjs';
import classnames from 'classnames';
import { useCopyToClipboard } from '../../hooks/useCopyToClipboard';
import styles from './index.module.scss';

interface Props {
  children: {
    props: {
      className: string | null;
      children: React.ReactNode;
    };
  };
}

const replaceLabelLanguages = (language: string) =>
  language
    .replace(/powershell/i, 'pwsh')
    .toUpperCase();


const CodeBox = ({ children: { props } }: Props): JSX.Element => {
  const [parsedCode, setParsedCode] = useState('');
  const [copied, copyText] = useCopyToClipboard();

  const className = props.className || 'text';

  // Language Matches in class
  const matches = className.match(/language-(?<lang>.*)/);

  // Language name from classname
  const language = matches?.groups?.lang || 'text';

  // Actual Code into a stringified format
  const stringCode = props.children?.toString() || '';

  const handleCopyCode = async (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    copyText(stringCode);
  };

  useEffect(() => {
    const prismLanguage = languages[language] || languages.text;
    setParsedCode(
      highlight(stringCode, prismLanguage, language)
    );
  }, []);

  return (
    <pre className={classnames(styles.pre, className)}>
      <div className={styles.top}>
        <span>{replaceLabelLanguages(language)}</span>
        <button type='button' onClick={handleCopyCode}>
          {copied ? 'copied' : 'copy'}
        </button>
      </div>
      <div
        className={styles.content}
        dangerouslySetInnerHTML={{ __html: parsedCode }}
      />
    </pre>
  );
};

export default CodeBox;