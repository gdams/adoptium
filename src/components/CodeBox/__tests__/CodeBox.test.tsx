import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import CodeBox from '../index';

Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn(),
  },
});

const navigatorClipboardSpy = vi.spyOn(navigator.clipboard, 'writeText');

afterEach(() => {
  vi.clearAllMocks();
});

describe('Codebox component', (): void => {
  it('renders correctly', (): void => {
    const textToCopy = <p>text to be copy</p>;
    const { container } = render(
      <CodeBox>
        {{
          props: {
            className: 'language-html',
            children: textToCopy,
          },
        }}
      </CodeBox>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly', async () => {
    const textToCopy = <p>text to be copy</p>;

    render(
      <CodeBox>
        {{
          props: {
            className: 'language-html',
            children: textToCopy,
          },
        }}
      </CodeBox>
    );

    navigatorClipboardSpy.mockImplementationOnce(() => Promise.resolve());

    const buttonElement = screen.getByText('copy');
    userEvent.click(buttonElement);

    await screen.findByText('copied');

    expect(navigatorClipboardSpy).toHaveBeenCalledTimes(1);
    expect(navigatorClipboardSpy).toHaveBeenCalledWith(textToCopy.toString());
  });
});