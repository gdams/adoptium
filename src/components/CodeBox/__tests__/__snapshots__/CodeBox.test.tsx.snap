// Vitest Snapshot v1

exports[`Codebox component > renders correctly 1`] = `
<div>
  <pre
    class="_pre_4cf629 language-html"
  >
    <div
      class="_top_4cf629"
    >
      <span>
        HTML
      </span>
      <button
        type="button"
      >
        copy
      </button>
    </div>
    <div
      class="_content_4cf629"
    >
      [object Object]
    </div>
  </pre>
</div>
`;
